### PR TITLE
Allow admins to specify TLS min version 1.3

### PR DESCRIPTION
**Pull request by maxsokolovsky**
## Problem
Admins cannot specify "tis-min-version" to be "1.3". Rancher validates the setting value and doesn't start.
 
## Solution
Change min version value validation and cipher list validation (the "tis-ciphers" setting), so that admins can start Rancher with a TLS min version 1.3 and its corresponding ciphers.

If admins set the version to 1.2, then we allow 1.2 and 1.3 ciphers.
If admins set the version to 1.3, then Rancher allows only 1.3 ciphers.

The default value is still 1.2 with its corresponding ciphers, so no change there.
 
### Automated Testing

* Test types added/modified:
    * Unit

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/tls/config.go [hunk 1]
```diff
@@ -45,54 +42,72 @@ var (
...
+	if version == tls.VersionTLS13 {
+		return validatedCiphersHelper(split, validCiphers13)
+	}
+	return validatedCiphersHelper(split, union(validCiphers10to12, validCiphers13))
+}
 
-	var TLSCiphers []uint16
-	for _, TLSCipher := range sliceTLSCiphers {
-		val, ok := tlsCipherSuites[strings.TrimSpace(TLSCipher)]
+func validatedCiphersHelper(ciphers []string, validCiphers map[string]uint16) ([]uint16, error) {
```
Comment (KevinJoiner) [inline pkg/tls/config.go:85]: nit:
I don't love calling things helper functions.
```suggestion
func getCipherValues(ciphers []string, validCiphers map[string]uint16) ([]uint16, error) {
```

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/tls/config.go [hunk 2]
```diff
@@ -3,25 +3,22 @@ package tls
 import (
 	"crypto/tls"
 	"fmt"
+	"sort"
 	"strings"
-
-	"github.com/rancher/rancher/pkg/settings"
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:8]: I did mention in the commit message that this is a refactoring of this part of the package.

There were a few issues before:
- unused package-level constants
- needless creation of a list of allowed versions/ciphers even when there is no validation error
- no pre-allocated slices
- not very good names
- coupling between this tls package and settings (these functions shouldn't care where the values are coming from)

I added tests first, then re-wrote this file. I agree that GH displays a lot of changes, and it might be hard to follow, but this is pretty much a rewrite of the file. Perhaps it's best to just review the new file on its own, without looking at how it was before.

File: pkg/tls/config.go [hunk 3]
```diff
@@ -45,54 +42,72 @@ var (
...
 		}
-		TLSCiphers = append(TLSCiphers, val)
 	}
-	return TLSCiphers, nil
+	return result
 }
 
-func getKeysFromMap(input map[string]uint16) []string {
-	var keys []string
+func keysFromMap(input map[string]uint16) []string {
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:109]: Because it's a refactoring of this part of the package, I thought it's OK to change unexported functions to give them better names.

File: pkg/tls/config_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			got, err := baseTLSConfig(test.minVersion, test.ciphers)
+			if err != nil && !test.errorExpected {
+				t.Fatalf("got an unexpected error: %v", err)
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config_test.go:86]: Yeah, I can see that. I usually stick with `testing` for simple enough tests like these.

File: pkg/tls/config.go [hunk 5]
```diff
@@ -45,54 +42,72 @@ var (
...
-		val, ok := tlsCipherSuites[strings.TrimSpace(TLSCipher)]
+func validatedCiphersHelper(ciphers []string, validCiphers map[string]uint16) ([]uint16, error) {
+	tlsCiphers := make([]uint16, 0, len(ciphers))
+	for _, cipher := range ciphers {
+		c, ok := validCiphers[strings.TrimSpace(cipher)]
 		if !ok {
-			return []uint16{}, fmt.Errorf("unsupported cipher [%s], must be one or more from: %s", TLSCipher, strings.Join(tlsCipherSuitesKeys, " "))
+			valid := keysFromMap(validCiphers)
+			sort.Strings(valid)
+			return nil, fmt.Errorf("unsupported cipher %s, must be one or more of:\n%s", cipher, strings.Join(valid, "\n"))
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:92]: Good point, will adjust! I initially added it for error cases as a readability aid, but it might be not pragmatic.

File: pkg/tls/config.go (hunk 1 above)
Comment (maxsokolovsky) [reply pkg/tls/config.go]: Yes, adjusted!

File: pkg/tls/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmurphy) [reply pkg/tls/config.go:8]: > I did mention in the commit message that this is a refactoring of this part of the package.

Yes, but you didn't say why. Could you put all those reasons in your commit message?

And better yet, could you separate the refactoring into a separate commit? This is widely considered best practice:

> In practice it often comes down to not mixing refactorings with new features. There are a few reasons for that:

> By mixing both things you increase the chance of having conflicts which are hard to resolve – not only for your code but also for code which is developed in parallel. Your code becomes harder to review: A Reviewer has to think about if a change is related to the new feature or just a rewrite of the existing code. Reverting your changes is becoming harder. Two atomic commits are much better: If there is an issue with the refactoring – Just revert it. If there is an issue with your new feature – Just revert it! A commit indicating “just did some cosmetic changes” might be a good hint to look further in the history.

https://programmerfriend.com/git-best-practices/

File: pkg/tls/config.go [hunk 6]
```diff
@@ -3,25 +3,22 @@ package tls
...
 		"1.2": tls.VersionTLS12,
+		// Deprecated.
 		"1.1": tls.VersionTLS11,
+		// Deprecated.
 		"1.0": tls.VersionTLS10,
 	}
 
 	// https://golang.org/pkg/crypto/tls/#pkg-constants
-	tlsCipherSuites = map[string]uint16{
+	validCiphers10to12 = map[string]uint16{
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply pkg/tls/config.go:21]: Yes but you could set the 1.3 versions apart by having a 1.3 for that variable, and leaving this one alone. See lines 149 and 201 in https://go.dev/src/crypto/tls/cipher_suites.go

File: pkg/tls/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmurphy) [reply pkg/tls/config.go:109]: It's not about exported vs unexported. My point is it's very subjective about what is better. IMO the old name was fine. If it doesn't enhance the functionality and isn't objectively more readable, it would be better to leave it alone. The more changes that are made, the harder it is to review and the greater the likelihood of introducing a bug.

**Review by cmurphy — CHANGES_REQUESTED** [review]
requesting changes just to ask that the "why" be added to the commit message

File: pkg/tls/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:21]: I think having those two slices with numbers in names is just more explicit. Besides, in the code listing you provided, I see variables like `supportedUpToTLS12` and `defaultCipherSuitesTLS13`, so I don't think our code is out of line with how that code lists different ciphers.

File: pkg/tls/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:109]: I definitely agree. I just think that in this case, it's permissible, since I planned on re-writing a relatively small piece of code, having added tests in advance as a safety net.

File: pkg/tls/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:8]: Also agree. Here, I thought it's just harder to rewrite, refactor, add new logic in separate commits. Normally, of course, I'd have separate commits for refactoring and tests.
I outlined the main reasons for the refactor above, just didn't include them in the commit message, because the overall change is small enough. And what would say? Not well-written functions?

[the remark this answers is not in the document]
Comment (cmurphy) [reply pkg/tls/config.go:8]: Could you just include those reasons in the commit message, that way someone wondering why it changed doesn't have to go to github and find this particular comment thread.

File: pkg/tls/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/tls/config.go:109]: As for good vs. bad names, I am not that obsessed with them, just thought its change is OK, given the opportunity to refactor.

Here is one example of a decent argument for verbs vs. nouns in method names: https://www.yegor256.com/2018/08/22/builders-and-manipulators.html

[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/tls/config.go:109]: @maxsokolovsky I understand your point and think it is a grey area on when to make stylistic changes, such as renaming variables or functions. If the change does not help maintainability, then it is not needed. This stops our history from being bloated with unnecessary name changes. And in this particular case, I don't think using a noun over a verb helps maintainability.

File: pkg/tls/config_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/tls/config_test.go:86]: You don't have to change just curious about why you prefer `testing`. 
I would prefer `testing` for simple cases if my repo didn't already import testify.

But for rancher tests I try and stick with testify for consistency, and the asserts flatten out the code and make them easier for myself to read.

File: pkg/tls/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/tls/config.go:21]: I don't mind the name change for this case because being explicit improves readability. Since you don't have to rely on code comments to explain version compatibility

**Review by KevinJoiner — APPROVED** [review]

**Review by cmurphy — APPROVED** [review]
